### DISCUSSION TITLE
Experience with the examples

### DISCUSSION THREAD

**Discussion by PhilippMolitor** [vortigont/MuiPlusPlus#4] [General]
So, I had a go with the examples.

I really had an easy time with the first (simple) example, though I have a bit of criticism:
It uses a lot of `new` keywords, which makes me a bit nervous about (de)allocation issues. Is there a way to make this safer, so it uses more stack allocated items?

The second example seems like you got some of the code of your ESP Iron firmware, and purged the more custom stuff out of it, still I find it too much to overview in one session. I would like to have an "scrolling list" example that is just as barebones as the other one.

Furthermore, I would like to have a logical overview that tells me what is possible with this library. All menu list types, all input types, something like that.

But so far, I feel very comfortable with this library approach and I will continue to use it for my project. Thanks for your effort in developing it, keep going!

**Comment by vortigont** [comment]
Hi!
Thanks for your feedback, much appreciated! This all still needs a lot of documenting and, possibly, reworking somewhere. I'm still not done yet with my Iron project that is driving this lib.

Regarding `new` - it's by design. All MuiItems are maintained as a list of shared pointers internally. There are two reasons for this:
 - those objects are all derived classes actually, so I have to use an indirection via pointer to base class anyway
 - it allows to dynamically add/remove menu objects on-demand. I use an approach to create/destruct it even when switching between menu sections to simplify my classes and keep it do a minimal set of functions

Do not worry about deallocation, those objects are destructed when container `MuiPluPlus` object is destroyed
if you want to keep an ownership of some of the menu item, you can attach it via shared pointer also, and keep a reference in your object to interact with it if required. I need to create an example of this, I think.
To add a bit of safety measures on object creation you can do something like this:
```cpp
auto quitbtn = new MuiItem_U8g2_ActionButton(...)
if (!quitbtn){
  // object creation failed for some reason, but it would probably throw exception anyway
  // do something
} else {
   muiplus.addMuippItem(quitbtn, root_page);
}
```
 
> I find it too much to overview in one session

I see, OK, I'll add intermediate example with just a scroll list of empty pages.


> All menu list types, all input types, something like that.

yes, it would take some efforts to describe those. All that I have for now you can see on those animated demo captures, I've recently added a second demo for power menu. A description for the objects is available within in-line doxygen ccomments in [muipp_u8g2.hpp](https://github.com/vortigont/MuiPlusPlus/blob/main/src/muipp_u8g2.hpp) along with methods description.

But just keep in mind that the whole idea of this lib is that you can create your own objects easily. Whatever you can draw with U8g2 you can use as a building brick for your menu. All the work on passing/taking key events for the object would be held by the MuiPlusPlus.

Cheers!

**Comment by vortigont** [comment]
@PhilippMolitor you can check this out [02_scrolling_menu](https://github.com/vortigont/MuiPlusPlus/tree/main/examples/02_scrolling_menu)
